packages/langium/test/grammar/type-system/inferred-types.spec.ts

```diff
@@ -0,0 +1,566 @@
+/******************************************************************************
+ * Copyright 2022 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import { AstTypes, createLangiumGrammarServices, Grammar, InterfaceType, isDataTypeRule, isParserRule, Property, PropertyType, UnionType } from '../../../src';
+import { collectInferredTypes } from '../../../src/grammar/type-system/inferred-types';
+import { parseHelper } from '../../../src/test';
+
+interface TypeContainer {
+    types: AstTypes;
+}
+
+function describeTypes(name: string, grammar: string, description: (this: TypeContainer) => void | Promise<void>): void {
+    describe(name, () => {
+        const container: TypeContainer = {
+            types: {
+                interfaces: [],
+                unions: []
+            }
+        };
+        beforeAll(async () => {
+            container.types = await getTypes(grammar);
+        });
+        description.call(container);
+    });
+}
+
+describeTypes('inferred types of simple grammars', `
+    A: name=ID value=NUMBER?;
+    B: name=FQN values+=NUMBER;
+    C: 'c' ref=[A];
+
+    FQN: ID;
+    terminal ID returns string: /string/;
+    terminal NUMBER returns number: /number/;
+`, function() {
```

Comment:
```suggestion
`, () => {
```

Reply:
I would love to, but there's a distinct difference between these approaches in how they handle the `this` parameter in TypeScript. `() => {}` expects to stay in the upper `this` context (in this case the file root context), while `function() {}` expects a dedicated `this: ...` parameter.

While both do the same thing in JS code, TypeScript infers the `this` type correctly when using `function() {}` but not when using `() => {}`.

Reply:
Why don't you pass the container as function argument?

Reply:
Actually, we don't even need the container, if we modify our `AstTypes` instance in place. Thanks for the hint 👍